---
title: '04: Differential activity analysis'
author: "felix.struebing@med.uni-muenchen.de"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

```{r}
library(tidyverse)
library(data.table)
library(Biostrings)
library(parallel)
library(tools)
library(Biobase)
library(pheatmap)
library(MPRAnalyze)
library(tidygenomics)
```

```{r}
pubfolder <- '/earth/public_data/'
datafolder <- '/earth/SNCA_MPRA/'
```

## Preparation
### Load data

```{r}
bcs <- fread(paste0(datafolder, 'LUHMES_barcodes/bcs_filtered_strandcorrected.tsv'), sep = '\t')
uniqueN(bcs$seq)
```

Separate tables are prepared containing DNA and RNA counts per barcode:

```{r}
bcs_RNA <- bcs[nacid == 'RNA', .(seq, group, replicate)]
bcs_RNA[, count := .N, by = .(seq, group, replicate)]
bcs_DNA <- bcs[nacid == 'gDNA', .(seq, group, replicate)]
bcs_DNA[, count := .N, by = .(seq, group, replicate)]
```

### Filtering

Only keep barcodes that are present in both the DNA and the RNA pool and remove any barcodes that are not present in at least 3 DNA replicates

```{r}
temp_DNA <- unique(bcs_DNA[, .(seq, group, replicate)]) %>% .[, count_reps := .N, by = c('seq', 'group')]
temp_DNA[, count_groups := .N, by = c('seq', 'replicate')]
```

For RNA, only accept barcodes with at least 3 counts > 0 in at least one group:

```{r}
temp_RNA <- unique(bcs_RNA[, .(seq, group, replicate)]) %>% .[, count_reps := .N, by = c('seq', 'group')]
temp_RNA[, count_groups := .N, by = c('seq', 'replicate')]
```

<!-- ```{r} -->
<!-- # table(temp_RNA$count_reps, temp_RNA$count_groups) -->
<!-- # table(temp_DNA$count_reps, temp_DNA$count_groups) -->
<!-- ``` -->


```{r}
temp_DNA <- temp_DNA[count_reps >= 3 & count_groups == 2,]
temp_RNA <- temp_RNA[count_reps >= 3]
myfil <- Reduce(intersect, list(temp_DNA$seq, bcs_RNA$seq, temp_RNA$seq))
uniqueN(myfil)
```

```{r}
RNA_mat_pre <- dcast(bcs_RNA[seq %in% myfil, ], seq ~ ..., value.var = 'count') 
DNA_mat_pre <- dcast(bcs_DNA[seq %in% myfil, ], seq ~ ..., value.var = 'count')
uniqueN(DNA_mat_pre$seq)
uniqueN(RNA_mat_pre$seq)
```

### Make count matrices

Count matrices suitable for `MPRAnalyze` are prepared:

```{r}
RNA_mat <- as.matrix(RNA_mat_pre[,-1])
rownames(RNA_mat) <- RNA_mat_pre$seq
DNA_mat <- as.matrix(DNA_mat_pre[,-1])
rownames(DNA_mat) <- DNA_mat_pre$seq
RNA_annot <- as.data.frame(unique(bcs_RNA[, .(group, replicate)]))
RNA_annot$group <- as.factor(RNA_annot$group)
rownames(RNA_annot) <- paste(RNA_annot$group, RNA_annot$replicate, sep = '_')
DNA_annot <- as.data.frame(unique(bcs_DNA[, .(group, replicate)]))
DNA_annot$group <- as.factor(DNA_annot$group)
rownames(DNA_annot) <- paste(DNA_annot$group, DNA_annot$replicate, sep = '_')
dim(RNA_mat) == dim(DNA_mat)
```

Results are saved: 

```{r eval = FALSE}
saveRDS(DNA_mat, paste0(datafolder, 'LUHMES_barcodes/DNA_count_matrix.rds'))
saveRDS(RNA_mat, paste0(datafolder, 'LUHMES_barcodes/RNA_count_matrix.rds'))
```

### Plot heatmap

```{r}
range <- min(cor(RNA_mat))
pheatmap(cor(DNA_mat), cluster_rows = FALSE, cluster_cols = FALSE,  breaks = seq(range, 1, length.out = 100), main = 'LUHMES gDNA counts: intercorrelations')
pheatmap(cor(RNA_mat), cluster_rows = FALSE, cluster_cols = FALSE, breaks = seq(range, 1, length.out = 100), main = 'LUHMES RNA counts: intercorrelations')
```

These heatmaps demonstrate that the DNA counts are highly related to each other and there is no obvious batch effect in transfection, whereas the RNA counts show clusters associated with our experimental design, i.e. the condition/group.

Next, the `MPRAnalyze` object is created and depth factors are estimated:

```{r eval = FALSE}
mpra <- MpraObject(dnaCounts = DNA_mat, rnaCounts = RNA_mat, dnaAnnot = DNA_annot, rnaAnnot = RNA_annot)
mpra <- estimateDepthFactors(mpra, lib.factor = c('group', 'replicate'), which.lib = 'both')
mpra <- analyzeQuantification(mpra, dnaDesign = ~ group, rnaDesign = ~ group, BPPARAM = BiocParallel::BatchtoolsParam(workers = 90L, cluster = BiocParallel::batchtoolsCluster()))
```

Finally, differential expression analysis is carried out using the groups as coefficient:

```{r eval = FALSE}
mpra <- analyzeComparative(mpra, dnaDesign = ~ group, rnaDesign = ~ group, reducedDesign = ~ 1, BPPARAM = BiocParallel::BatchtoolsParam(workers = 90L, cluster = BiocParallel::batchtoolsCluster()))
saveRDS(mpra, paste0(datafolder, 'LUHMES_barcodes/mpranalyze_dataset_bygroup.rds'))
```

<!-- This function would only analyze barcodes where each DNA counts has an associated RNA counts and vice versa: -->
<!-- ```{r} -->
<!-- mpra <- analyzeComparative(mpra, rnaDesign = ~ group, reducedDesign = ~ 1, mode = 'scale', BPPARAM = BiocParallel::BatchtoolsParam(workers = 90L, cluster = BiocParallel::batchtoolsCluster())) -->
<!-- ``` -->

## Analysis

### Likelihood-ratio test

A likelihood ratio test is carried out with `MPRAnalyze`:

```{r}
lrt <- testLrt(mpra)
```

Barcodes with little power (NA in logFC) are removed: 

```{r}
res <- lrt %>% 
    dplyr::filter(!is.na(logFC)) %>% 
    arrange(fdr) %>% 
    mutate(seq = rownames(.))
```

Number of significant results:

```{r}
nrow(res %>% dplyr::filter(fdr < 0.1 & abs(logFC) > 0.5))
nrow(res %>% dplyr::filter(fdr < 0.1 & abs(logFC) > 1))
nrow(res %>% dplyr::filter(fdr < 0.1 & abs(logFC) > 2))
```

### Barcode correspondence

The correspondence table is loaded and filtered for valid barcodes that survived the likelihood ratio test:

```{r}
correspondence <- fread(paste0(datafolder, 'blat_results/correspondence_table_unfiltered.tsv'), sep = '\t', nThread = 80)
correspondence <- correspondence[seq %in% res$seq, .(qname, chr = chr_R1, sstart_R1, sstart_R2, send_R1, send_R2, seq)]
```

Number of barcodes in LUHMES MPRA experiment:

```{r}
uniqueN(res$seq)
```

Number of barcodes that are also present in the (filtered) correspondence table:

```{r}
uniqueN(correspondence$seq)
```

The strand information is extracted:

```{r}
# correspondence$seq <- as.character(reverseComplement(DNAStringSet(correspondence$seq))) # this was done before, kept for legacy
correspondence <- correspondence[, `:=` (strand = ifelse(sstart_R2 > sstart_R1, '+', '-'), width_R1 = abs(send_R1 - sstart_R1), width_R2 = abs(send_R2 - sstart_R2))][width_R1 <= 151 & width_R2 <= 151, ]
```

Since only the start and end positions of the RE were sequenced, contiguous intervals are defined:

```{r}
correspondence[, `:=` (start = min(sstart_R1, sstart_R2, send_R1, send_R2), end = max(sstart_R1, sstart_R2, send_R1, send_R2)), by = qname]
correspondence <- correspondence[, .(qname, chr, start, end, strand, seq)]
correspondence[, locus := paste(chr, start, end, strand, sep = '_')]
```


<!-- ```{r} -->
<!-- correspondence[, n_locus := .N, by = locus] -->
<!-- correspondence[, n_barcode := .N, by = seq] -->
<!-- correspondence[, n_bc_per_locus := .N, by = .(locus, seq)] -->
<!-- ``` -->

Barcodes aligning to the same locus are counted, and for each barcode, the top 5 loci are kept:

```{r}
corr_filt <- correspondence[, .(seq, locus)]
corr_filt <- corr_filt[, .(n = .N), by = c('seq', 'locus')]
corr_filt <- setorder(setDT(corr_filt), seq, -n)[, head(.SD, 5), keyby = seq]
```

If any of the four top loci per barcode are in each other's vicinity (defined by a maximum distance of `max_distance` bps defined below), they are considered the same locus and receive the same `cluster_id`.

```{r}
mycorr <- corr_filt %>% as_tibble() %>% 
    arrange(-n) %>% 
    dplyr::mutate(strand = str_sub(locus, start = -1L)) %>% 
    separate(locus, into = c('chr', 'start', 'end', sep = '_', extra = 'drop')) %>% 
    dplyr::select(seq, chr, start, end, strand, n) %>% 
    dplyr::mutate(start = as.numeric(start), end = as.numeric(end)) %>% 
    arrange(-n) %>% group_by(seq) %>% slice_max(n, n = 4, with_ties = FALSE) %>% 
    group_by(seq) %>% 
    mutate(rank = rank(n, ties.method = 'first')) %>% 
    genome_cluster(by = c('chr', 'start', 'end'), max_distance = 50) %>% 
    group_by(seq, cluster_id) %>% 
    mutate(coverage = sum(n))
uniqueN(mycorr$cluster_id)
uniqueN(mycorr$seq)
```

For barcodes and loci that are less than `max_distance` apart and share the same `cluster_id`, they are considered the same locus if the second top_ranking alignment within the `max_distance` window makes up less than 25% of total counts, and the locus with the most counts is considered the "true" locus. In that case, their counts are summed up to yield coverage information.

```{r}
mycorr_filt <- mycorr %>%
    mutate(locus = paste(chr, start, end, strand, sep = '_')) %>% 
    group_by(seq) %>% 
    dplyr::mutate(accept = case_when(
        rank == 4 & n[2] / n[1] <= 0.25 & n[1] >= 10 ~ 'yes',
        rank == 4 & cluster_id[1] == cluster_id[2] & n[1] >= 5~ 'yes',
        TRUE ~ 'no'
    )) %>% 
    dplyr::filter(accept == 'yes') %>% 
    dplyr::select(seq, locus, chr, start, end, strand, coverage)
```

Some loci will be represented by more than one barcode. Number of unique barcodes and loci left after filtering procedure:

```{r}
uniqueN(mycorr_filt$seq)
uniqueN(mycorr_filt$locus)
```

Coverage plot of correspondence table (i.e. the number of reads supporting a given barcode-locus correspondence):

```{r}
covplot <- data.frame(index = seq(from = 1, to = 1000, by = 5))
covplot$unique_seq <- sapply(covplot$index, function(x) {
    uniqueN(mycorr_filt[mycorr_filt$coverage > x,]$seq)
})
covplot$unique_locus <- sapply(covplot$index, function(x) {
    uniqueN(mycorr_filt[mycorr_filt$coverage > x,]$locus)
})
# covplot$in_luhmes <- sapply(covplot$index, function(x) {
#     length(mycorr_filt[mycorr_filt$coverage > x,]$seq %in% res$seq)
# })
covplot %>% 
    ggplot(aes(x = index)) +
    geom_line(aes(y = unique_seq, color = 'red')) +
    geom_line(aes(y = unique_locus, color = 'dodgerblue1')) +
    # geom_line(aes(y = in_luhmes, color = 'black')) +
    labs(x = 'Coverage filter cutoff', y = 'n', title = 'Unique barcodes and loci per coverage cutoff') +
    scale_colour_manual(name = 'Color', 
         values =c('red'='red','dodgerblue1'='dodgerblue1'), labels = c('Unique barcodes','Unique loci')) +
    scale_x_sqrt(breaks = c(0, 5, 10, 20, 50, 100, 200, 500, 1000))
    
```

Curves are starting to near each other when considering barcodes with a coverage of approx. 50


### Final filtering for valid barcodes

```{r}
covcut = 50
table(res$seq %in% mycorr_filt$seq)
res_fil <- res %>% right_join(., mycorr_filt %>% dplyr::filter(coverage >= covcut), by = 'seq') %>% 
    select(seq, chr, start, end, strand, coverage, fdr, pval, logFC) %>% 
    mutate(significant = ifelse(fdr <= 0.1 & abs(logFC) > 0.5, 'yes', 'no'))
table(res_fil$seq %in% mycorr_filt$seq)
```








################################### break 220708
















Overlap with design:

```{r}
res_fil.gr <- res_fil %>% makeGRangesFromDataFrame(., keep.extra.columns = TRUE)
```

```{r}
rois <- readRDS('01_export/rois.rds')
on_target <- table(res_fil.gr %over% rois)
on_target[2] / length(res_fil.gr)
```

```{r}
res_fil.gr$ontarget <- res_fil.gr %over% rois
table(res_fil.gr$significant, res_fil.gr$ontarget)
```

Can the low on-target percentage be explained with overlaps to repeat regions?

```{r}
repeats.gr <- read_delim('/earth/public_data/hg38_repeatMasker_UCSC', delim = '\t') %>% 
    dplyr::select('genoName', 'genoStart', 'genoEnd', 'strand', 'repName', 'repFamily') %>% 
    makeGRangesFromDataFrame(., seqnames.field = 'genoName', start.field = 'genoStart', end.field = 'genoEnd', keep.extra.columns = TRUE)
hits <- findOverlaps(res_fil.gr, repeats.gr, select = 'arbitrary')
res_fil.gr$has_repeat <- hits
res_fil.gr$has_repeat <- ifelse(is.na(res_fil.gr$has_repeat), 'no', 'yes')
```

Repeats by type:

```{r}
all_hits <- findOverlaps(res_fil.gr, repeats.gr, select = 'all')
repeat_hits <- res_fil.gr[queryHits(all_hits)]
repeat_hits$repeat_family <- repeats.gr[subjectHits(all_hits)]$repFamily
```

```{r}
table(repeat_hits$repeat_family)
as.data.frame(repeat_hits) %>% 
    group_by(repeat_family) %>% 
    add_tally() %>% 
    dplyr::filter(n >= 10) %>% 
    summarise(n_repeats = sum(n)) %>% 
    ggplot(aes(x = repeat_family, y = n_repeats, fill = significant)) +
    geom_col(position = 'dodge')
```


```{r}
library(ggsignif)
rt <- addmargins(table(res_fil.gr$has_repeat, res_fil.gr$ontarget), c(1,2))
rt
fisher.test(table(res_fil.gr$has_repeat, res_fil.gr$ontarget))
as.data.frame(table(res_fil.gr$has_repeat, res_fil.gr$ontarget)) %>% 
    mutate(ontarget = Var2, has_repeat = Var1) %>% 
    ggplot(aes(x = has_repeat, fill = ontarget, y = Freq)) +
    geom_col(position = 'fill', color = 'black') +
    theme_light(base_size = 14) +
    labs(y = 'Cumulative frequency') +
    geom_signif(xmin = 'no', xmax = 'yes', annotation = c('***'), y_position = 1.2, tip_length = 0.0001) +
    scale_y_continuous(labels = scales::percent)
```

```{r}
addmargins(table(res_fil.gr$significant, res_fil.gr$ontarget), c(1,2))
fisher.test(table(res_fil.gr$significant, res_fil.gr$ontarget), alternative = 'less')
as.data.frame(table(res_fil.gr$significant, res_fil.gr$ontarget)) %>% 
    mutate(significant = Var1, ontarget = Var2) %>% 
    ggplot(aes(x = significant, fill = ontarget, y = Freq)) +
    geom_col(position = 'fill', color = 'black') +
    theme_light(base_size = 14) +
    labs(y = 'Cumulative frequency') +
    geom_signif(xmin = 'no', xmax = 'yes', annotation = c('p = 0.1514'), y_position = 1.2, tip_length = 0.0001) +
    scale_y_continuous(labels = scales::percent)
```


How to read this contingency table: no/yes refers to whether or not a barcode overlaps a repeat region, FALSE/TRUE refers to whether or not the barcode/locus was included in the design.
Therefore, `r rt[1,1] / rt[1,3] * 100` % of off-target loci had no repeat, whereas `r rt[2,1]/ rt[2,3] * 100` % of off-target loci did have a repeat overlap, a result that is highly significant when using fisher's exact test:

```{r}
fisher.test(table(res_fil.gr$has_repeat, res_fil.gr$ontarget))
fisher.test(table(res_fil.gr$significant, res_fil.gr$ontarget), alternative = 'less')
```

```{r}
summary(lm(data = mcols(res_fil.gr), formula = logFC ~ has_repeat))
```


Since more off-target loci/ROIs/baits have repeats, this could explain the somewhat low on-target rate (this was expected though). 

<!-- ```{r} -->
<!-- res_fil.gr$has_repeat <- 'no' -->
<!-- res_fil.gr$repeatfam <- NA -->
<!-- res_fil.gr$rep_width <- 0 -->
<!-- res_fil.gr[queryHits(hits)]$has_repeat <- 'yes' -->
<!-- res_fil.gr[queryHits(hits)]$repeatfam <- repeats.gr[subjectHits(hits)]$repFamily -->
<!-- res_fil.gr[queryHits(hits)]$rep_width <- width(pintersect(res_fil.gr[queryHits(hits)],repeats.gr[subjectHits(hits)])) -->
<!-- ``` -->

Export the results:

```{r}
res_fil <- as.data.frame(res_fil.gr) %>% 
    mutate(locus = paste0(seqnames, ':', start, '-', end)) %>% 
    dplyr::select(seq, locus, strand, fdr, logFC, everything())
```

### Plots and stats

NOTE 2022-05-12 also plot colors by GC content, has_repeat, ontarget, coverage etc! to rule out batch effect

```{r}
res_fil %>% 
    ggplot(aes(x = logFC, y = -log10(pval), color = significant)) +
    geom_point(alpha = 0.7) +
    labs(x = 'fold change [log2]', y = '-log10(pvalue)', title = 'LUHMES MPRA: volcano plot [SNCA vs. GFP]') +
    geom_hline(yintercept = -log10(
        res_fil %>% ungroup() %>% dplyr::filter(significant == 'yes') %>% dplyr::slice_max(fdr, n = 1) %>% dplyr::pull(pval)), linetype = 'dashed', alpha = 0.3) +
    theme_light(base_size = 14)
```

```{r}
res_fil %>% 
    ggplot(aes(x = logFC, y = -log10(pval), color = ontarget)) +
    geom_point(alpha = 0.7) +
    labs(x = 'fold change [log2]', y = '-log10(pvalue)', title = 'LUHMES MPRA: volcano plot [SNCA vs. GFP]') +
    geom_hline(yintercept = -log10(
        res_fil %>% ungroup() %>% dplyr::filter(significant == 'yes') %>% dplyr::slice_max(fdr, n = 1) %>% dplyr::pull(pval)), linetype = 'dashed', alpha = 0.3) +
    theme_light(base_size = 14)
```


```{r}
res_fil %>% 
    ggplot(aes(x = logFC, y = -log10(pval), color = log2(coverage))) +
    geom_point(alpha = 0.8, size = 2) +
    scale_color_gradient(high = 'tomato3', low = 'steelblue2') +
    labs(x = 'fold change [log2]', y = '-log10(pvalue)', title = 'LUHMES MPRA: volcano plot [SNCA vs. GFP]') +
    geom_hline(yintercept = -log10(
        res_fil %>% ungroup() %>% dplyr::filter(significant == 'yes') %>% dplyr::slice_max(fdr, n = 1) %>% dplyr::pull(pval)), linetype = 'dashed', alpha = 0.3) +
    theme_light(base_size = 14)
```


```{r}
res_fil %>% 
    ggplot(aes(x = logFC, y = -log10(pval), color = has_repeat)) +
    geom_point(alpha = 0.7, size = 2) +
    labs(x = 'fold change [log2]', y = '-log10(pvalue)', title = 'LUHMES MPRA: volcano plot [SNCA vs. GFP]') +
    geom_hline(yintercept = -log10(
        res_fil %>% ungroup() %>% dplyr::filter(significant == 'yes') %>% dplyr::slice_max(fdr, n = 1) %>% dplyr::pull(pval)), linetype = 'dashed', alpha = 0.3) +
    theme_light(base_size = 14)
```


```{r}
res_fil %>% 
    ggplot(aes(x = logFC, y = -log10(pval), color = scale(width))) +
    geom_point(alpha = 1, size = 2) +
    scale_color_gradient2(low = 'tomato3', high = 'steelblue2', mid = 'grey90') +
    labs(x = 'fold change [log2]', y = '-log10(pvalue)', title = 'LUHMES MPRA: volcano plot [SNCA vs. GFP]') +
    geom_hline(yintercept = -log10(
        res_fil %>% ungroup() %>% dplyr::filter(significant == 'yes') %>% dplyr::slice_max(fdr, n = 1) %>% dplyr::pull(pval)), linetype = 'dashed', alpha = 0.3) +
    theme_light(base_size = 14)
```

```{r}
saveRDS(res_fil, '14_export/filtered_result_table.rds')
```



## ADDENDUM - CONFIDENTIAL SANITY CHECK - not for share

The clusters that are visible to the right in the volcano plot seem unusual. Batch effect?
MANUAL inspection of the clusters: Seems legit- a negative FC has suppressed RNA counts in SNCA and counts of a few dozens to hundreds in GFP, despite equal DNA counts in both (hundreds). A positive FC is associated with higher expression in SNCA.

For the bottom-right cluster:

```{r}
DNA_mat <- readRDS('13_export/DNA_count_matrix.rds')
RNA_mat <- readRDS('13_export/RNA_count_matrix.rds')
test_seqs <- res_fil %>% dplyr::filter(-log10(pval) > 1 & -log10(pval) < 5 & abs(logFC) > 3) %>% dplyr::pull(seq)  %>% head()
```

```{r}
mypca <- prcomp(t(RNA_mat))
summary(mypca)

mypca_tib <- mypca$x[,1:2] %>% 
    as_tibble() 
mypca_tib$sample <- rownames(mypca$x) 
mypca_tib %>% 
    mutate(group = str_sub(sample, end = -3L)) %>% 
    ggplot(aes(x = PC1, y = PC2, color = group)) +
    geom_point(size = 4) +
    theme_light(base_size = 14) +
    labs(title = 'PCA of RNA count matrix')
    
```

```{r}
mypca <- prcomp(t(DNA_mat))
summary(mypca)

mypca_tib <- mypca$x[,1:2] %>% 
    as_tibble() 
mypca_tib$sample <- rownames(mypca$x) 
mypca_tib %>% 
    mutate(group = str_sub(sample, end = -3L)) %>% 
    ggplot(aes(x = PC1, y = PC2, color = group)) +
    geom_point(size = 4) +
    theme_light(base_size = 14) +
    labs(title = 'PCA of DNA count matrix')
    
```

```{r}
DNA_mat[test_seqs,]
```

```{r}
RNA_mat[test_seqs,]
```

For the top-right cluster:

```{r}
test_seqs <- res_fil %>% dplyr::filter(-log10(pval) > 10 & -log10(pval) < 15 & abs(logFC) > 3) %>% dplyr::pull(seq)  %>% head()
```

```{r}
DNA_mat[test_seqs,]
```

```{r}
RNA_mat[test_seqs,]
```


## Results table

Hint: Entries in the "locus" column can just be copied and pasted to the UCSC genome browser (hg38) for manual inspection.
```{r}
DT::datatable(res_fil %>% arrange(fdr))
```


